### PR TITLE
Ignore "code style" commits in the blame view

### PR DESCRIPTION
**Pull request by rr-it**
### Description:

Introduce `.git-blame-ignore-revs` to ignore commits doing only code style or other automatic changes.

This declutters the git blame view and by that eases  the retrace of code changes.

Activate via:
+ GitHub: automatically
+ local: `git config --local blame.ignoreRevsFile .git-blame-ignore-revs`

See:
+ https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view
+ https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by michalkleiner** [comment]
Generally makes sense to me, thanks for the contribution @rr-it.

I wonder whether we should add a sentence about this to documentation and also who and when would determine and based on what criteria which commits should be added to this.

And as a side note — I never had any issue using git blame where I would cry for this to be used even when larger code shifts were done, as long as the commit messages are well written and descriptive enough. But I'm not opposed to adding this.

**Comment by sgiehl** [comment]
@michalkleiner I had similar thoughts on that. The "problem" I see is that I'm unsure if this will be maintained properly...

**Comment by rr-it** [comment]
Maintenance of the ignore list is kind of cumbersome: the commit id is only generated on merge/commit. So an extra commit to `.git-blame-ignore-revs` is needed after merging/commiting the PR itself.

You might check for 'candidates' once in a while:

```bash
git log --abbrev=40 --pretty=format:"%h" --shortstat | grep -Pzo "[a-z0-9]{40}\n[^\n]+[0-9]{3,} files changed[^\n]*[0-9]{4,} insertions[^\n]*[0-9]{4,} deletions[^\n]*\n?"
```

<details>

<summary>Candidates for `.git-blame-ignore-revs` (click to open)</summary>

065efa10af6cc13f142ed157c1dcdf4f1b3f378b
 258 files changed, 2362 insertions(+), 2411 deletions(-)
aea8f07de8ea3ce6ce39d3a217af5ec6a957db96
 112 files changed, 4829 insertions(+), 4378 deletions(-)
6ceadb6ba2e7ff95baa91ecb18899f52180c49c5
 288 files changed, 16321 insertions(+), 4640 deletions(-)
6ca414ae3307edd6e37162986d0f926f9b301a72
 122 files changed, 1115 insertions(+), 1050 deletions(-)
466840c33205c7e8393d583d24a92ac314f40913
 160 files changed, 2057 insertions(+), 2053 deletions(-)
ff812d1faeab4d7f712289b53883b14a4e6910cb
 238 files changed, 19466 insertions(+), 25157 deletions(-)
7c8227bb111af98f766fc48c2f6141b1115c41ab
 176 files changed, 1760 insertions(+), 1009 deletions(-)
f5eda650a61a5c1aaa9073e71bd7b5643bfe52af
 126 files changed, 1481 insertions(+), 1007 deletions(-)
db1f2bbc968a3d7e5e4da5ce1c482625548742b6
 106 files changed, 2413 insertions(+), 2233 deletions(-)
40afe8b101537e1ca4c28d345b756a22cbaeac21
 126 files changed, 2495 insertions(+), 1787 deletions(-)
43c21c99ef0451772a52a449729f832bfad8e82e
 111 files changed, 6197 insertions(+), 1510 deletions(-)
bd118d12701009d72576fed7dd4e5bdf45ab3368
 116 files changed, 1699 insertions(+), 1276 deletions(-)
2ac2bc1aeaf8efce6bb9af92506311da99e1757f
 109 files changed, 99394 insertions(+), 1463 deletions(-)
627c9e6484d0dc3819c32d2ee05bcb010bcc9982
 148 files changed, 4496 insertions(+), 4183 deletions(-)
92353c0d25260abfcfa3a1423e027b35a155c6ac
 259 files changed, 4194 insertions(+), 4028 deletions(-)
c6bb7314071c18cf0ef54a4e35d3531314e68792
 269 files changed, 4603 insertions(+), 13209 deletions(-)
de5ae85e7a11f7bbd962256c890a32188c40d000
 190 files changed, 8676 insertions(+), 6976 deletions(-)
d0aaab95f1ef6d8b748b0e8d54f811a366864249
 1061 files changed, 1967 insertions(+), 174975 deletions(-)
115527353a9e75e01aa4d263408956ae45403bea
 2496 files changed, 2904 insertions(+), 2933 deletions(-)
6936b93cba5150e0eaa879aebf3662b3e279d045
 439 files changed, 41070 insertions(+), 11197 deletions(-)
dfafa5720f14e7a7419a25e0d59ad86806b6113c
 110 files changed, 1338 insertions(+), 1354 deletions(-)
93aef4865cfdee9fcfa5acc9ff1950459a0af42e
 472 files changed, 2393 insertions(+), 2309 deletions(-)
9e34ecd4ff43d19dee4b4e881545302d07887bb7
 180 files changed, 1819 insertions(+), 1707 deletions(-)
3f26e785f015d30d0aeea66aaf7484111b0dbfa9
 456 files changed, 74369 insertions(+), 1185 deletions(-)
b16a791aa3650d85af829156129c2bd44c7cb075
 2027 files changed, 2027 insertions(+), 2027 deletions(-)
76e60590f486f33a38e7ea08bf941fa57ac904c5
 596 files changed, 7392 insertions(+), 5320 deletions(-)
808780232c38b036bb903e3e434eede7f25e00a9
 392 files changed, 1744 insertions(+), 1121 deletions(-)
2e006803ee17a8d1a992085c6425eddaa84a25f5
 113 files changed, 4000 insertions(+), 1249 deletions(-)
509ac2f1085e74c6a44253248f9f2fc3ac54d337
 159 files changed, 5739 insertions(+), 5683 deletions(-)
34c4a857126cd936ff1e9b1a4fb590306be1c879
 201 files changed, 145008 insertions(+), 4356 deletions(-)
0ad551add46e9758523b7068278059d68e9152c0
 446 files changed, 1628 insertions(+), 1521 deletions(-)
ae0d7185cee2f73f053c0bb546e9054e6fe183f1
 560 files changed, 2074 insertions(+), 1681 deletions(-)
07adf0ac04dfb8f35be41d9add1855b6e646f8db
 1033 files changed, 6668 insertions(+), 6640 deletions(-)
9af4e95aa976f3a6533e95b776b5298f73e5f916
 293 files changed, 6386 insertions(+), 2697 deletions(-)
660df51720386077f0a1433d86bcb96ef002de35
 169 files changed, 12290 insertions(+), 7213 deletions(-)
6a97e43ff69795c53c1b8dff63fc6a5cbdba8ca8
 299 files changed, 4328 insertions(+), 3891 deletions(-)
a0f0eb8fa0cc86e14afeabde80d2a21a7e0122d9
 682 files changed, 1301 insertions(+), 1278 deletions(-)
57e33fffc158ad086c6bf7c8fc5ce5e0a6263bb2
 767 files changed, 1189 insertions(+), 1182 deletions(-)
5a2a0e4dd41f64a01e3427739d7594881d7ac2d7
 128 files changed, 2210 insertions(+), 1368 deletions(-)
587cc39e0362719332d410b7a4d5ddcc68788eeb
 234 files changed, 15991 insertions(+), 1672 deletions(-)
e59e8c57d613ee8a42951b6525256dc0ceb0b061
 275 files changed, 29842 insertions(+), 10697 deletions(-)
c3abe93f76ae2c160fb96d6d63b9a2400b55447b
 428 files changed, 1111 insertions(+), 1355 deletions(-)
0c9c30b731ccbacf47e154b9f7a590af49e3d799
 846 files changed, 61717 insertions(+), 22166 deletions(-)
252148a742fa0681066d16a2aba30d429ef45e25
 115 files changed, 3991 insertions(+), 1168 deletions(-)
b52ae4e7e488e0474d67c54578e1d6c1aa066bff
 197 files changed, 8841 insertions(+), 4157 deletions(-)
2bae9d90f860f3ccc5598420eb978cafd598451e
 1294 files changed, 11797 insertions(+), 7747 deletions(-)
626612546502728ad9c108c36b019d1968d86088
 423 files changed, 2284 insertions(+), 2178 deletions(-)
1432d00d57602f7cba7cc9b161effcb12a70b600
 530 files changed, 1064 insertions(+), 1101 deletions(-)
10d482f1d3deb4c717fbb87f523428196634a326
 180 files changed, 2171 insertions(+), 2804 deletions(-)
6b836717d6391adf891e2fd60f8787f033d726b8
 124 files changed, 7875 insertions(+), 7834 deletions(-)
9ba8f216fd7856ce5fef06bf82ecb8f8a2e7e630
 448 files changed, 31657 insertions(+), 15247 deletions(-)
d17d6b37afb10809b80a15290a868b58fd5da706
 161 files changed, 1100 insertions(+), 5420 deletions(-)
647adaff61a16ed46a18ef0cdf2266a6ad1e288e
 555 files changed, 2370 insertions(+), 1234 deletions(-)
0bbbdc851366a1b5cd7179a7de313caa655a6fda
 110 files changed, 3261 insertions(+), 3054 deletions(-)
a67a4d5e9bfab1a6e3834e38a3fc2739e9375212
 164 files changed, 3568 insertions(+), 1318 deletions(-)
806bedba1a43d7d78b687d1d23a6b861a2b70035
 233 files changed, 1098 insertions(+), 1098 deletions(-)
a2fb93968cd111c25b3112f3283dcfd685e4b7de
 929 files changed, 33501 insertions(+), 20920 deletions(-)
20c7801cf3ba4db2a4b9728e6043488a7d147b5d
 1057 files changed, 138190 insertions(+), 39074 deletions(-)
3e134db7cf87e28b37037b650b3378fe9deeb475
 167 files changed, 1350 insertions(+), 1178 deletions(-)
a00487b0b841c4b15463b591c7f62176c4b84d15
 228 files changed, 2632 insertions(+), 1541 deletions(-)
1415b5e1b88adc16f352c786c2681356da9ba082
 365 files changed, 3741 insertions(+), 1721 deletions(-)
c1573b9becddcaf3156294d11b3ee6945d2ec08a
 1676 files changed, 74339 insertions(+), 64210 deletions(-)
d9adcfe6169c6c10059a670f2ed984908eb4e105
 401 files changed, 5631 insertions(+), 5632 deletions(-)
26e73549e33322b2ef00940aa3efc3e6284298a6
 1679 files changed, 64063 insertions(+), 74174 deletions(-)
62ba8dfb44a4b6f8fbe5345dd0c0ea5a4d52033b
 1679 files changed, 74174 insertions(+), 64063 deletions(-)
f52c1e83aef857b6cf073f1f9fb00c6982f8d7d2
 131 files changed, 3552 insertions(+), 1881 deletions(-)
bd09242078458761f9905285b5ca349172fdd149
 102 files changed, 6043 insertions(+), 1496 deletions(-)
81ccb29b440e98ed9147e84781df0fce34ee6542
 147 files changed, 1295 insertions(+), 1230 deletions(-)
1fa8da9b963e99e08c829fe6491e8ccf2d1054e2
 170 files changed, 1199 insertions(+), 1181 deletions(-)
ba582598731bda1e238e3dfd2b9e00389c32166d
 100 files changed, 85063 insertions(+), 81448 deletions(-)
a96ec40e983a8b64f4c25def541a519ff3670e6a
 294 files changed, 3837 insertions(+), 3263 deletions(-)
5104d94f3b2250f766b9c520e2da8da9b4cab2e9
 358 files changed, 3091 insertions(+), 2621 deletions(-)
aca8ec33a85e0a70b818b6227145041717d7a69e
 321 files changed, 3328 insertions(+), 2757 deletions(-)
6e38973acc12e9f7fa2d02edee3f8bbbfdbb9694
 271 files changed, 2039 insertions(+), 2037 deletions(-)
03b4953f008c1063e6d7166143ba844e8c6e89cc
 220 files changed, 1365 insertions(+), 1365 deletions(-)
9b2c0a7a450fff3b634f8119c8003ae1d20b97d0
 220 files changed, 1711 insertions(+), 1476 deletions(-)
287dcbbe9eecd9d6e4c5cd89a5a2fa7fe14bb09a
 814 files changed, 50108 insertions(+), 93015 deletions(-)
ae4b03163792f0b6e933933e5d37df87dc3fd566
 854 files changed, 121350 insertions(+), 121446 deletions(-)
884e54cf378bffdea2dd5360af8a260b25352746
 101 files changed, 1716 insertions(+), 1635 deletions(-)
7e36c6db68ece71d4a270e4ccc361854df494140
 409 files changed, 1263 insertions(+), 1237 deletions(-)
7191afd6d6bb443708886bb013a4155271ce40d3
 257 files changed, 12440 insertions(+), 1516 deletions(-)
63a4254ca8f6bea0189e82c8890bdcbcbeeb3fba
 384 files changed, 1619 insertions(+), 1637 deletions(-)
1378f79297baa84590738839ec11cef90ac21bc4
 377 files changed, 18615 insertions(+), 2349 deletions(-)
b3e81145346875719956918533bc4b13f49add49
 127 files changed, 1121 insertions(+), 1083 deletions(-)
ada0ccf4ecfd94428d0b2a74ea85607722989aaa
 296 files changed, 39321 insertions(+), 5531 deletions(-)
16aab5b2e15377a74c6b3345de82291b90bfa8b5
 166 files changed, 4474 insertions(+), 3386 deletions(-)
66ea73f5cecb08df8cac7706c65018bb6719278d
 301 files changed, 17705 insertions(+), 17700 deletions(-)
0c3ef698010708ba0484aabd5a01db3b62f47365
 100 files changed, 3265 insertions(+), 2293 deletions(-)
42b52b6d8a88b3fa4c4f3978c4e7bf00b1eac778
 243 files changed, 1031 insertions(+), 29822 deletions(-)
acc4fd8ebc685c5c9afd314b6ea224a911bfd405
 156 files changed, 21615 insertions(+), 10866 deletions(-)
d4f405b6550da50a4cb67b5b78a05f2117f56f2a
 801 files changed, 71109 insertions(+), 71109 deletions(-)

</details>

**Comment by michalkleiner** [comment]
I think I'm leaning towards accepting this as it is now since locally you need to invoke it explicitly (or knowingly configure your local git to respect the ignore list), and within GitHub you get a notification it's in effect and it can be disabled for the blame view.

We can think of ways how to manage this later on if needed.

What do you reckon @bx80?

**Comment by bx80** [comment]
I think this could be useful for hiding things like mass PSR code improvements which touch lots of files but aren't functionally meaningful, something which we may be doing more of in the future. I use git blame quite a lot to understand the evolution of the code so anything that decreases noise there would be welcome. 

Agreed it would be good add a few words into the dev docs, even if it's just a copy of the notes in this ticket :+1:

File: .git-blame-ignore-revs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# many typo fixes in the code thanks to codespell (#15730)
+2c5c72e2da4c015efa78069c9c1b407fc55a6483
+# Update doc blocks with new name (#15857)
+115527353a9e75e01aa4d263408956ae45403bea
+# Typofixes (#19805)
+d3de8d64f477d14bf10d5cd1d17521f0c19950ed
+# [PHPCS] force new line at end of php file (#20715)
+593437a3fd35381c13fe5cf7fabae5767dc3a519
+# [PHPCS] Enable PSR1 coding standard (#20724)
+aa2674a8241d5e4e55a2f7fbb74292e06eab8780
```
Comment (sgiehl) [inline .git-blame-ignore-revs:25]: There's a new commit we can add now...

```suggestion
aa2674a8241d5e4e55a2f7fbb74292e06eab8780
[PHPCS] disallow using tabs for intending
d40e11f6decbe84834ed5f33406189055e6989b8
```

File: .git-blame-ignore-revs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+[PHPCS] disallow using tabs for intending
```
Comment (sgiehl) [inline .git-blame-ignore-revs]:
```suggestion
# [PHPCS] disallow using tabs for intending
```

**Review by sgiehl — APPROVED** [review]
I'll merge this one now. Even if we don't manage to maintain that continuously, it useful and can be used and extended when ever needed though.

**Pull request merged**